enhancements/okd-on-fedora-coreos.md

```diff
@@ -0,0 +1,76 @@
+authors:
+  - "@LorbusChris"
+reviewers:
+  - "@smarterclayton"
+  - "@miabbott"
+  - "@ashcrow"
+  - "@runcom"
+  - "@darkmuggle"
+  - "@vrutkovs"
+approvers:
+creation-date: 2019-10-21
+last-updated: 2019-10-21
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:  
+replaces:
+superseded-by:
+---
+
+# OKD-on-Fedora-CoreOS
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+ 1. This proposal potentially requires addition of non-Koji-built RPMs from OpenShift's Prow CI instance to Fedora CoreOS composes. Can we do this? If yes, how do we do this?
+ 2. What is the specific list of operators that will go into OKD-on-FCoS?
+
+ Please see the Graduation Criteria section for more context on both open questions.
+
+## Summary
+
+A proposal to create components of an OKD variant that runs atop Fedora CoreOS and to use OpenShift's Prow instance to do so.
+
+As part of the version 4 effort, the OKD working group has agreed to target Fedora CoreOS (FCoS) as the primary base operating system for builds of version 4 of the OKD Community Distribution that is powered by the OpenShift codebase.
+
+This document proposes to incrementally add Prow CI jobs to create Fedora-based artifacts for use in OKD version 4.
+
+## Motivation
+
+Delivering continuous builds of version 4 of the OKD community variant, based on Fedora CoreOS. 
+
+### Goals
+
+The broader goal is to create automated builds of all components needed for the OKD-on-FCoS effort, with a preference on running jobs in Fedora containers, and exclusively using RPMs from official Fedora repositories for Fedora CoreOS composes. Container artifacts from UBI based CI builds for OCP may be re-used, however those artifacts that come from RHEL container based builds may not be re-used to build OKD.
+
+### Non-Goals
+
+The migration to Ignition Spec v3 is not part of this proposal.
+
+## Proposal
+
+In order to create automated builds of OKD-on-FCoS, OpenShift's Prow instance shall be used to run jobs to build all required components/artifacts in Fedora containers. The definitions of these jobs will live side by side with the already existing definitions for OCP jobs in the `openshift/release` repository, on the respective operators' repos. Operators that will not go into OKD can be disregarded from this effort.
+
+### Graduation Criteria
+
+Specifically, success of this proposal will be measured by the following:
+- RPMs that are dependencies of OpenShift's codebase in the base OS are built continuously in a Fedora environment: `cri-o`, `openshift-hyperkube`, `openshift-clients`, `machine-config-daemon`
+  - **Open Question**: How are builds promoted from Prow CI to FCoS release composes?
```

Comment:
Will those RPMs be part of FCoS?  That implies we have one FCoS release per OKD.  I would assume we would pull FCoS streams into OKD layers and simply boot from FCoS AMIs

Reply:
From discussions with @bgilbert and @miabbott I was under the impression we'd add those RPMs to the base OS manifest, because we need them on the bootstrap host already.

Can you clarify what you mean with OKD layers? the machine-os-content?

Reply:
We could change the bootstrap host to pivot as well to break that dependency.

In other words, this path FCOS is pure bootimage, and we just have an "OKDFCOS" oscontainer/`machine-os-content` which contains the kubelet.

Reply:
Having the bootstrap pivot too would be another way to fix https://github.com/openshift/installer/pull/2532

Reply:
If we only have to produce our machine-os-content container, and not worry about the deps on the bootstrap host (i.e. in FCoS), that would be very nice indeed.

Reply:
It is a MASSIVE problem if we are not pivoting the bootstrap node.  That means we don't have hermetic installs based on release image, because the chosen AMI could break.

That is a serious, serious bug.

Reply:
> Is it correct that we'd then just need the MCD (as RPM) on the bootstrap host?

I think as part of this, the installer would extract the MCD binary from the release image, or alternatively run the MCD container image via podman, and do the pivot that way.  In other words, the installer would only require `podman` on the initial bootstrap host.

Reply:
Filed https://github.com/openshift/installer/issues/2542 to track this as a dependency and ensure we have consensus from the installer team.

Reply:
Filed https://github.com/openshift/installer/issues/2542 to track this as a dependency and ensure we have consensus from the installer team.